tools/omnissm/api/register/register.go

```diff
@@ -0,0 +1,289 @@
+	InstanceType     string `json:"instanceType"`
+}
+
+// GetManagedID Retrieve instance ssm managed instance id
+func (i *InstanceIdentity) GetManagedID() string {
+	// Not Intended to be cryptographically secure, just a partition / lookup key
+	// we only get s
+	ident := strings.Join([]string{i.AccountID, i.InstanceID}, "-")
+	h := sha1.New()
+	h.Write([]byte(ident))
+	bid := h.Sum(nil)
+	return fmt.Sprintf("%x", bid)[0:17]
+}
+
+// GetRegistration fetch instance registration from db
+func (i *InstanceIdentity) GetRegistration() (*InstanceRegistration, error) {
+	registration := &InstanceRegistration{}
+
+	params := &dynamodb.GetItemInput{
+		TableName: aws.String(RegistrationTable),
+		AttributesToGet: []string{
+			"mid", "ActivationId", "ActivationCode",
+		},
+		Key: map[string]dynamodb.AttributeValue{
+			"mid": {
+				S: aws.String(i.GetManagedID()),
+			},
+		},
+	}
+
+	getRequest := dbClient.GetItemRequest(params)
+	getResult, err := getRequest.Send()
+	if err != nil {
+		return registration, err
+	}
+	err = dynamodbattribute.UnmarshalMap(getResult.Item, registration)
+	if err != nil {
+		return registration, err
+	}
+	return registration, nil
+}
+
+// RegisterInstance Create SSM activation for instance and store
+func (i *InstanceIdentity) RegisterInstance() (*InstanceRegistration, error) {
+
+	registration := &InstanceRegistration{}
+
+	activateParams := &ssm.CreateActivationInput{
+		IamRole: aws.String(SSMInstanceRole),
+		Description: aws.String(
+			strings.Join([]string{i.AccountID, i.InstanceID}, "-")),
+	}
+	activateReq := ssmClient.CreateActivationRequest(activateParams)
+	activateResult, err := activateReq.Send()
+
+	if err != nil {
+		return registration, err
+	}
+	registration.ActivationCode = *activateResult.ActivationCode
+	registration.ActivationID = *activateResult.ActivationId
+	registration.ManagedID = i.GetManagedID()
+
+	regRecord, err := dynamodbattribute.MarshalMap(registration)
+	insertParams := &dynamodb.PutItemInput{
+		Item:      regRecord,
+		TableName: aws.String(RegistrationTable),
+	}
+
+	insertRequest := dbClient.PutItemRequest(insertParams)
+	insertResult, err := insertRequest.Send()
+
+	if err != nil {
+		fmt.Println("Put Registration Error", insertResult, err)
+		return registration, err
+	}
+	return registration, nil
+}
+
+// InstanceRegistration Minimal
+type InstanceRegistration struct {
+	ActivationCode string
+	ActivationID   string `json:"ActivationId"`
+	ManagedID      string `json:"mid"`
+}
+
+func validateRequest(requestBody string) (InstanceIdentity, events.APIGatewayProxyResponse) {
+
+	var regRequest RegistrationRequest
+	var identity InstanceIdentity
+
+	err := json.Unmarshal([]byte(requestBody), &regRequest)
+	if err != nil {
+		response := map[string]string{
+			"error":   "invalid-request",
+			"message": "malformed json",
+		}
+		body, _ := json.Marshal(response)
+		return identity, events.APIGatewayProxyResponse{Body: string(body), StatusCode: 400}
```

Comment:
This might make easier to read/write code if you return a pointer to `events.APIGatewayProxyResponse` instead of having to initialize a struct for each return.  If nothing else I would make use of named return values to cut down on struct literals being returned.